### PR TITLE
Roles

### PR DESCRIPTION
**Pull request by markivancho**

### CODE REVIEW

File: spec/factories/users.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FactoryGirl.define do
+
+  factory :user do
+    email { FFaker::Internet.email }
```
Comment (AndrewMatsievsky) [inline spec/factories/users.rb:4]: I think you mistyped Faker

Comment (markivancho) [reply spec/factories/users.rb:4]: No, it's slightly different [gem](https://github.com/ffaker/ffaker).

File: Gemfile [hunk 2]
```diff
@@ -40,8 +40,12 @@ gem 'autoprefixer-rails'
 gem 'omniauth'
 gem 'omniauth-google-oauth2'
 gem 'figaro'
-
-
+gem 'pundit'
+gem 'rspec'
```
Comment (jizak) [inline Gemfile]: all gems related to tests should go to `test` group.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      redirect_to users_path
+    else
+      flash[:error] = "Something went wrong"
+      redirect_to users_path
+    end
+  end
+
+
+  def user_params
+      params.require(:user).permit(:email, :role)
```
Comment (jizak) [inline app/controllers/users_controller.rb]: styles

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    authorize @user
+  end
+
+  def show
+    @user = User.find(params[:id])
+    authorize @user
+  end
+
+  def destroy
+    @user = User.find(params[:id])
```
Comment (jizak) [inline app/controllers/users_controller.rb]: we have couple actions where we are look for a user and `authorize` him.
It makes sense to move it to separate private method and do `before_action`

File: app/policies/user_policy.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class UserPolicy
+  attr_reader :current_user, :model
+
+  def initialize(current_user, model)
+    @current_user = current_user
+    @user = model
+  end
+
+  def index?
+    @current_user == @user
```
Comment (jizak) [inline app/policies/user_policy.rb]: Lets do that only admin can access users CRUDS.
Current User will have a separate `Profile` page.

File: app/policies/user_policy.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  end
+  
+  def show?
+    @current_user.admin? || @current_user == @user
```
Comment (jizak) [inline app/policies/user_policy.rb]: only `admin` for CRUD.

File: app/views/users/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<%= form_for(@user) do |f| %>
+  <div class="form-inputs">
+    <p><%= f.text_field :email %></p>
+    <p><%= f.select :role, ["admin", "member"]%></p>
```
Comment (jizak) [inline app/views/users/edit.html.erb]: You should get this from `enum` methods. If in future we add another role you won't need to add it manually. (potential bug).
`User.roles.keys` - something like that. Play with it in controller.

File: app/views/users/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<div class="well">
+<% @users.each do |user| %>
+  <p>
+  <%= user.email %>
```
Comment (jizak) [inline app/views/users/index.html.erb]: intendation.

File: app/views/users/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<div class="well">
+<% @users.each do |user| %>
+  <p>
+  <%= user.email %> 
+  <% if current_user.admin? %>
```
Comment (jizak) [inline app/views/users/index.html.erb]: looks like this html can be separated into partial.
Also, it is a bad practice to have mixed hash styles: `:class => 'btn btn-xs btn-danger', method: :delete`

File: config/database.yml [hunk 10]
```diff
@@ -58,6 +58,8 @@ development:
 test:
   <<: *default
   database: myapp_test
+  username: postgres
```
Comment (jizak) [inline config/database.yml]: To be honest your app should work without those creds. The same in development.

File: spec/features/signin_spec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+feature "signing_in" do
```
Comment (jizak) [inline spec/features/signin_spec.rb:3]: these are integration tests. You didn't need to do them :)

File: spec/policies/user_policy_spec.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    it 'allows owner to see profile' do
+      expect(subject).to permit(current_user, current_user)
+    end
+
+    it 'allows an admin to see profile' do
+      expect(subject).to permit(admin, other_user)
+    end
+  end
+
+  permissions :update? do
```
Comment (jizak) [inline spec/policies/user_policy_spec.rb:24]: don't forget to update tests after changing policies. Actually, you should run your tests before each PR.

File: app/views/users/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<%= form_for(@user) do |f| %>
+  <div class="form-inputs">
+    <p><%= f.text_field :email %></p>
+    <p><%= f.select(:role, User.roles.keys.map {|role| [role.titleize,role]})%></p>
```
Comment (jizak) [inline app/views/users/edit.html.erb]: I know that I am boring but please use styleguides :) `User.roles.keys.map{ |role| [role.titleize,role] }) %>`

File: config/database.yml [hunk 14]
```diff
@@ -17,6 +17,8 @@
 default: &default
   adapter: postgresql
   encoding: unicode
+  username: postgres
```
Comment (jizak) [inline config/database.yml:20]: do you need it here? Maybe you can't see problem here. But imagine if multiple developers work on the same project. They all need to change `database.yml` everytime they need to run an app.

File: spec/controllers/users_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      get :edit, params: { id: user.id }
+      expect(response).to render_template("edit")
+    end
+
+  end
+
+  describe "DELETE #destroy" do
+    it "deletes the user" do
+      get :destroy, params: { id: other_user.id }
+      expect(flash[:success]).to eql("User deleted")
```
Comment (jizak) [inline spec/controllers/users_controller_spec.rb]: this check is good but not enough. It is better to check if the user was actually destroyed.
As I remember ActiveRecord has `destroyed?` method. ex: `other_user.destroyed?`

File: spec/controllers/users_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+  end
+
+  describe "PATCH #update" do
+    context "with good data" do
+      it "updates an user" do
+        patch :update, params: { id: user.id, user: FactoryGirl.attributes_for(:user, email: "email@email.com") }
+        user.reload
+        expect(flash[:success]).to eql("User updated")
```
Comment (jizak) [inline spec/controllers/users_controller_spec.rb]: Again it is better to check the email was updated.

**Review by jizak — APPROVED** [review]

**Pull request merged**